composer/trainer/mosaic_fsdp.py

```diff
@@ -0,0 +1,312 @@
+    param_dtype = None
+    reduce_dtype = None
+    buffer_dtype = None
+    if isinstance(mixed_precision, dict):
+        param_dtype = mixed_precision.get('param_dtype', None)
+        if param_dtype is not None:
+            param_dtype = get_torch_dtype(param_dtype)
+        reduce_dtype = mixed_precision.get('reduce_dtype', None)
+        if reduce_dtype is not None:
+            reduce_dtype = get_torch_dtype(reduce_dtype)
+        buffer_dtype = mixed_precision.get('buffer_dtype', None)
+        if buffer_dtype is not None:
+            buffer_dtype = get_torch_dtype(buffer_dtype)
+    elif isinstance(mixed_precision, str):
+        mixed_precision = mixed_precision.upper()
+        if mixed_precision == 'FULL':
+            pass
+        elif mixed_precision == 'DEFAULT':
+            reduce_dtype = get_torch_dtype(precision)
+            buffer_dtype = torch.float32
+        elif mixed_precision == 'PURE':
+            param_dtype = get_torch_dtype(precision)
+            reduce_dtype = get_torch_dtype(precision)
+            buffer_dtype = get_torch_dtype(precision)
+        else:
+            raise ValueError(f'Unable to interpret mixed_precision={mixed_precision}')
+    else:
+        raise ValueError(f'Unable to interpret mixed_precision={mixed_precision}')
+
+    mixed_precision = MixedPrecision(
+        param_dtype=param_dtype,
+        reduce_dtype=reduce_dtype,
+        buffer_dtype=buffer_dtype,
+        keep_low_precision_grads=keep_low_precision_grads,
+    )
+
+    return mixed_precision, param_dtype, reduce_dtype, buffer_dtype
+
+
+def _get_cpu_offload(cpu_offload=False):
+    """Helper fn for configuring cpu_offload."""
+    cpu_offload = CPUOffload(offload_params=True) if cpu_offload else None
+    if cpu_offload is not None:
+        raise ValueError('FSDP CPU Offload not supported yet.')
+    return cpu_offload
+
+
+def _get_process_group(pg, process_group_cache=None):
+    """Helper fn for configuring and/or retrieving process groups."""
+    warnings.warn(f'Instantiating FSDP with custom process groups is an experimental feature.')
+
+    # Return regular process_groups as is, no cacheing
+    if pg is None or isinstance(pg, ProcessGroup):
+        return pg
+
+    world_size = dist.get_world_size()
+
+    # Handle special str process_group cases
+    if pg == 'self':
+        pg = 'set1'
+        warnings.warn(f"Converting process_group='self' to process_group='{pg}'")
+    elif pg == 'node':
+        pg = f'set{world_size}'
+        warnings.warn(f"Converting process_group='node' to process_group='{pg}'")
+    elif pg == 'local_rank_across_nodes':
+        pg = f'mod{world_size}'
+        warnings.warn(f"Converting process_group='local_rank_across_nodes' to process_group='{pg}'")
+
+    # Handle str and Union[List[int], Tuple[int]] process_group cases
+    if isinstance(pg, str) and pg.startswith('set'):
+        k = int(pg.strip('set'))
+        world_size = dist.get_world_size()
+        if world_size % k:
```

Comment:
Don't we need an equality check `== 0` here?

Reply:
if `world_size` is divisible by `k` then `world_size % k == 0` and we do not need to raise the error.
with an `== 0` check you'd also need to add a not
```
if not world_size % k == 0:
```

right?

Reply:
`world_size % k != 0` would be a bit clearer I think, just so I don't have to think about which case is which